redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,709 @@
+use crate::aio::ConnectionLike;
+use crate::cluster_async::{ClusterConnInner, Connect, Core, InnerCore};
+use crate::cluster_routing::SlotAddr;
+use crate::cluster_topology::SLOT_SIZE;
+use crate::{cmd, from_redis_value, Cmd, ErrorKind, RedisError, RedisResult, Value};
+use async_trait::async_trait;
+use std::sync::Arc;
+
+/// This module contains the implementation of scanning operations in a Redis cluster.
+///
+/// The `ClusterScanArgs` struct represents the arguments for a cluster scan operation,
+/// including the scan state reference, match pattern, count, and object type.
+///
+/// The `ScanStateCursor` struct is a wrapper for managing the state of a scan operation in a cluster.
+/// It holds a reference to the scan state and provides methods for accessing the state.
+///
+/// The `ClusterInScan` trait defines the methods for interacting with a Redis cluster during scanning,
+/// including retrieving address information, refreshing slot mapping, and routing commands to specific addresss.
+///
+/// The `ScanState` struct represents the state of a scan operation in a Redis cluster.
+/// It holds information about the current scan state, including the cursor position, scanned slots map,
+/// address being scanned, and address's epoch.
+
+const BITS_PER_U64: usize = u64::BITS as usize;
+const NUM_OF_SLOTS: usize = SLOT_SIZE as usize;
+const BITS_ARRAY_SIZE: usize = NUM_OF_SLOTS / BITS_PER_U64;
+const END_OF_SCAN: u16 = NUM_OF_SLOTS as u16 + 1;
+type SlotsBitsArray = [u64; BITS_ARRAY_SIZE];
+
+#[derive(PartialEq, Debug, Clone)]
+pub(crate) struct ScanState {
+    // the real cursor in the scan operation
+    cursor: u64,
+    // a map of the slots that have been scanned
+    scanned_slots_map: SlotsBitsArray,
+    // the address that is being scanned currently, based on the next slot set to 0 in the scanned_slots_map, and the address "own" the slot
+    // in the SlotMap
+    pub(crate) address_in_scan: String,
+    // epoch represent the version of the address, when a failover happens or slots migrate in the epoch will be updated to +1
+    address_epoch: u64,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ClusterScanArgs {
+    pub(crate) scan_state_cursor: ScanStateCursor,
+    match_pattern: Option<String>,
+    count: Option<usize>,
+    object_type: Option<ObjectType>,
+}
+
+#[derive(Debug, Clone)]
+/// Represents the type of an object in Redis.
+pub enum ObjectType {
+    /// Represents a string object in Redis.
+    STRING,
```

Comment:
Nit: Rust convention is to use CamelCase enumerators `String`, `List`, `Set` etc